Self-Hosted E2E Conformance Tests Fail kubectl Executions

/kind feature

I am running self-hosted E2E conformance tests in a pod on every freshly deployed cluster (similar as in https://github.com/ozdanborne/k8s-e2e-containerized). e2e.test is started without any host, kubeconfig configuration and picks up the InClusterConfig correctly. All conformance tests pass nicely, except from kubectl executions.

kubectl gets executed by e2e.test as `kubectl --server=http://127.0.0.1:8080 ...` and therefore fails on connection refused. In the container running kubectl without --server, will work as expected, since I have service accounts set up correctly.

Looking into code, for some reason defaultHost is set to http://127.0.0.1:8080 ( https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/test_context.go#L33 ). That value is not used within e2e.test as it seems, but is passed on to kubectl, only.

Omitting `--server=http://127.0.0.1:8080` would let the kubectl run with no problem, why is it set to that value (default proxy port is 8001)?

Currently I am mitigating by pre-exectuting `kubectl proxy --port=8080 &`
before running e2e.test.

```
Summarizing 24 Failures:

[Fail] [k8s.io] SchedulerPredicates [Serial] [BeforeEach] validates that NodeSelector is respected if not matching [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/scheduling/predicates.go:84

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl run job [It] should create a job from an image when restart is OnFailure [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Pods Extended [k8s.io] Delete Grace Period [It] should be submitted and removed [Conformance] [Flaky] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pods.go:162

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl patch [It] should add annotations for pods in rc [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] SchedulerPredicates [Serial] [BeforeEach] validates resource limits of pods that are allowed to run [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/scheduling/predicates.go:84

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl api-versions [It] should check if v1 is in available api versions [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl version [It] should check is all data is printed [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl run --rm job [It] should create a job from an image, then delete the job [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl describe [It] should check if kubectl describe prints relevant information for rc and pods [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl run rc [It] should create an rc from an image [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl run default [It] should create an rc or deployment from an image [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl label [BeforeEach] should update the label on a resource [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl logs [BeforeEach] should be able to retrieve and filter logs [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl expose [It] should create services for rc [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Guestbook application [It] should create and stop a working application [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Update Demo [It] should do a rolling update of a replication controller [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Update Demo [It] should create and stop a replication controller [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Update Demo [It] should scale a replication controller [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Proxy server [It] should support proxy with --port 0 [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:1482

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl replace [It] should update a single-container pod's image [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl run deployment [It] should create a deployment from an image [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl rolling-update [It] should support rolling-update to same image [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

[Fail] [k8s.io] SchedulerPredicates [Serial] [BeforeEach] validates that NodeSelector is respected if matching [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/scheduling/predicates.go:84

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl run pod [It] should create a pod from an image when restart is Never [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2124

Ran 149 of 652 Specs in 2600.345 seconds
FAIL! -- 125 Passed | 24 Failed | 0 Pending | 503 Skipped --- FAIL: TestE2E (2600.38s)
FAIL
```

--- a/test/e2e/framework/test_context.go
+++ b/test/e2e/framework/test_context.go
@@ -19,12 +19,16 @@ package framework
 import (
 	"flag"
 	"fmt"
+	"io/ioutil"
 	"os"
 	"time"
 
+	"github.com/golang/glog"
 	"github.com/onsi/ginkgo/config"
 	"github.com/spf13/viper"
+	restclient "k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 	"k8s.io/kubernetes/pkg/cloudprovider"
 	"k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig"
 	"k8s.io/kubernetes/pkg/kubemark"
@@ -296,12 +300,61 @@ func ViperizeFlags() {
 	AfterReadingAllFlags(&TestContext)
 }
 
+func createKubeConfig(clientCfg *restclient.Config) *clientcmdapi.Config {
+	clusterNick := "cluster"
+	userNick := "user"
+	contextNick := "context"
+
+	config := clientcmdapi.NewConfig()
+
+	credentials := clientcmdapi.NewAuthInfo()
+	credentials.Token = clientCfg.BearerToken
+	credentials.ClientCertificate = clientCfg.TLSClientConfig.CertFile
+	if len(credentials.ClientCertificate) == 0 {
+		credentials.ClientCertificateData = clientCfg.TLSClientConfig.CertData
+	}
+	credentials.ClientKey = clientCfg.TLSClientConfig.KeyFile
+	if len(credentials.ClientKey) == 0 {
+		credentials.ClientKeyData = clientCfg.TLSClientConfig.KeyData
+	}
+	config.AuthInfos[userNick] = credentials
+
+	cluster := clientcmdapi.NewCluster()
+	cluster.Server = clientCfg.Host
+	cluster.CertificateAuthority = clientCfg.CAFile
+	if len(cluster.CertificateAuthority) == 0 {
+		cluster.CertificateAuthorityData = clientCfg.CAData
+	}
+	cluster.InsecureSkipTLSVerify = clientCfg.Insecure
+	config.Clusters[clusterNick] = cluster
+
+	context := clientcmdapi.NewContext()
+	context.Cluster = clusterNick
+	context.AuthInfo = userNick
+	config.Contexts[contextNick] = context
+	config.CurrentContext = contextNick
+
+	return config
+}
+
 // AfterReadingAllFlags makes changes to the context after all flags
 // have been read.
 func AfterReadingAllFlags(t *TestContextType) {
 	// Only set a default host if one won't be supplied via kubeconfig
 	if len(t.Host) == 0 && len(t.KubeConfig) == 0 {
-		t.Host = defaultHost
+		// Check if we can use the in-cluster config
+		if clusterConfig, err := restclient.InClusterConfig(); err == nil {
+			if tempFile, err := ioutil.TempFile(os.TempDir(), "kubeconfig-"); err == nil {
+				kubeConfig := createKubeConfig(clusterConfig)
+				clientcmd.WriteToFile(*kubeConfig, tempFile.Name())
+				t.KubeConfig = tempFile.Name()
+				glog.Infof("Using a temporary kubeconfig file from in-cluster config : %s", tempFile.Name())
+			}
+		}
+		if len(t.KubeConfig) == 0 {
+			glog.Warningf("Unable to find in-cluster config, using default host : %s", defaultHost)
+			t.Host = defaultHost
+		}
 	}
 	// Reset the cluster IP range flag to CLUSTER_IP_RANGE env var, if defined.
 	if clusterIPRange := os.Getenv("CLUSTER_IP_RANGE"); clusterIPRange != "" {

